### PR TITLE
Fix IMAP SCAN on selected mailbox crasher

### PR DESCRIPTION
**Pull request by ksmurchison**
imapd.c: needs an open mailbox for index_scan(), imapd_index is released before every command is parsed, so we need to reopen the mailbox

### CODE REVIEW

**Review by elliefm — CHANGES_REQUESTED** [review]
The fix looks fine on a code read, but can you please add a test to prove it (and to prevent us breaking it again in the future)?  Thanks

**Review by rsto — CHANGES_REQUESTED** [review]
Agree with @elliefm, code looks good but please add a test.

**Comment by ksmurchison** [comment]
Actually, SCAN was broken worse than just the crasher.  It always returned nothing due to a logic error in the ACL test.  And then it returned EXISTS, RECENT, and FLAGS responses fro every mailbox that is scanned.  All issues have been fixed and tested - SCAN with/without a previously SELECTed mailbox

**Review by elliefm — APPROVED** [review]
Looks good, thanks

**Review by rsto — APPROVED** [review]

**Comment by rsto** [comment]
I fixed a memory leak in the new code and pushed an update.

File: cassandane/Cassandane/Cyrus/List.pm [hunk 1]
```diff
@@ -2329,4 +2329,55 @@ sub test_list_return_subscribed_trash_nochildren
...
+    my $store = $self->{store};
+    my $imaptalk = $store->get_client();
+
+    $self->setup_mailbox_structure($imaptalk, [
+        [ 'create' => [qw( INBOX.a INBOX.b INBOX.c )] ],
+    ]);
+
+    xlog $self, "listing...";
+    my $res = $imaptalk->list("", "*");
+warn Dumper($res);
```
Comment (elliefm) [inline cassandane/Cassandane/Cyrus/List.pm:2346]: can you remove this warning though please :)

Comment (rsto) [reply cassandane/Cassandane/Cyrus/List.pm]: done

**Pull request merged**